File: java/direct-policy/src/main/java/org/nhindirect/policy/impl/machine/StackMachineEntry.java
package org.nhindirect.policy.impl.machine;

import org.nhindirect.policy.Opcode;
import org.nhindirect.policy.PolicyOperator;
import org.nhindirect.policy.PolicyValue;

public class StackMachineEntry implements Opcode
{
	public static enum EntryType
	{
		OPERATOR,
		
		VALUE;
	}
	
	protected final EntryType entryType;
	protected final PolicyOperator operator;
	protected final PolicyValue<?> value;
	
	public StackMachineEntry(PolicyOperator operator)
	{
		this.operator = operator;
		this.entryType = EntryType.OPERATOR;
		this.value = null;
	}
	
	public StackMachineEntry(PolicyValue<?> value)
	{
		this.operator = null;
		this.entryType = EntryType.VALUE;
		this.value = value;
	}
	
	public PolicyValue<?> getValue()
	{
		return value;
	}
	
	public PolicyOperator getOperator()
	{
		return operator;
	}
	
	public EntryType getEntryType()
	{
		return entryType;
	}
	
	@Override
	public String toString()
	{
		final StringBuilder builder = new StringBuilder("Entry Type: ")
				.append(getEntryType()).append("\r\n");
		switch(getEntryType())
		{
			case OPERATOR:
				builder.append("Operator: " + getOperator().getOperatorText());
				break;
			case VALUE:
				builder.append("Value: " + getValue().getPolicyValue());
				break;	
		}
		
		return builder.toString();
	}
}

File: csharp/policy/Machine/StackMachineEntry.cs
using System;
using System.Security.Cryptography.X509Certificates;
using Health.Direct.Policy.Operators;
using System.Text;

namespace Health.Direct.Policy.Machine
{
    //public class LiteralMachineEntry<T> : StackMachineEntry<T> 
    //{
    //    public LiteralMachineEntry(object policyValue)
    //        : base(EntryType.Value, policyValue)
    //    {
    //    }
    //}

    
    //public class ReferenceMachineEntry<T> : StackMachineEntry<T> 
    //{
    //    public ReferenceMachineEntry(object policyValue)
    //        : base(EntryType.Value, policyValue)
    //    {
    //    }
    //}


    

    public class StackMachineEntry<T> : IOpCode
    {
        readonly OperatorBase m_policyOperator;
        readonly EntryType m_entryType;
        readonly object m_policyValue;
        readonly Type m_opCodeType;

        public EntryType EntryType { get { return m_entryType; } }
        public OperatorBase PolicyOperator { get { return m_policyOperator; } }
        public object PolicyValue { get { return m_policyValue; } }

        public StackMachineEntry(OperatorBase policyOperator)
        {
            m_policyOperator = policyOperator;
            m_entryType = EntryType.Operator;
            m_policyValue = null;
            m_opCodeType = typeof(T);
        }

        public StackMachineEntry(EntryType entryType,  object policyValue)
        {
            m_entryType = entryType;
            m_policyValue = policyValue;
            m_opCodeType = typeof (T);
        }
       
        
        protected StackMachineEntry(){}

        
        public OperatorBase GetOperator()
        {
            return PolicyOperator;
        }
        

        public override string ToString()
        {
            StringBuilder builder = new StringBuilder("Entry Type: ")
                    .Append(EntryType).Append("\r\n");
            switch (EntryType)
            {
                case EntryType.Operator:
                    builder.Append("Operator: " + GetOperator().GetOperatorText() + "\r\n");
                    break;
                case EntryType.Value:
                    builder.Append("Value: " + PolicyValue + "\r\n");
                    break;
            }

            return builder.ToString();
        }
    }

    public enum EntryType
    {
        Operator,
        Value
    }
}
